### PR TITLE
feat: improved unlocking display

### PR DESCRIPTION
**Pull request by aulneau**
### What this does

This PR resolves https://github.com/blockstack/explorer/issues/362 and https://github.com/blockstack/explorer/issues/361.

There is some logic on the API endpoint i've made to determine if an address has vesting data to fetch its true balance from the `v2/account` node endpoint, and the components will use that value when applicable. Note: this does not apply to if a user searches for the address in the search bar. The updated, correct information is only displayed on the Address page. Once the fixes to the API go into effect, then everything should be the same.

The lockup data is static, and should not change. [You can view the data for it here.](https://raw.githubusercontent.com/blockstack/explorer/feat/better-unlocking/src/pages/api/vesting/lockups.json)

| Address view | Unlocking schedule view |
| ------------- | ----------------------- |
| Screen Shot 2021-02-12 at 12 06 11 PM      | Screen Shot 2021-02-12 at 12 06 16 PM       |

## Staging & example addresses

- [SP31AJM5WD8N9WB3HRCN5M2MA5QB14RF4ZYGEY9D7](https://explorer-git-feat-better-unlocking.blockstack.now.sh/address/SP31AJM5WD8N9WB3HRCN5M2MA5QB14RF4ZYGEY9D7)
- [SP2WPBNG1XR3HG95R91RVBYN4SADTBYZFFAV7Y4BQ](https://explorer-git-feat-better-unlocking.blockstack.now.sh/address/SP2WPBNG1XR3HG95R91RVBYN4SADTBYZFFAV7Y4BQ)
- [SPH4BRKR7J2BTHRZE2CPVYCSKPDXWP9EWSV61QBA](https://explorer-git-feat-better-unlocking.blockstack.now.sh/address/SPH4BRKR7J2BTHRZE2CPVYCSKPDXWP9EWSV61QBA)
- [SP2Y1FXTR9BJH29VKQ35FKFRXC40149RB5T9TVN6J](https://explorer-git-feat-better-unlocking.blockstack.now.sh/address/SP2Y1FXTR9BJH29VKQ35FKFRXC40149RB5T9TVN6J)

### todo
- [x] Mobile view for unlocking schedule

### CODE REVIEW

File: src/common/hooks/use-unlocking-state.ts [hunk 1]
```diff
@@ -0,0 +1,8 @@
+import { useRecoilState } from 'recoil';
+import { unlockingState } from '@store/unlocking';
+
+export const useUnlockingState = () => {
+  const [state, setState] = useRecoilState(unlockingState);
+
+  return [state, setState];
```
Comment (kyranjamie) [inline src/common/hooks/use-unlocking-state.ts:7]:
```suggestion
  return useRecoilState(unlockingState);
```
?

File: src/components/stacking.tsx [hunk 2]
```diff
@@ -57,7 +57,10 @@ export const StackingPercentage = ({ balances, stackingBlock }: any) => {
                 </Title>
               </Flex>
               <Box textAlign="center">
-                <Caption mb="base-tight">Stacking for ~{blocksUntilUnlocked} more blocks</Caption>
+                <Caption mb="base-tight">
+                  ~{blocksUntilUnlocked.toLocaleString(undefined, { maximumFractionDigits: 2 })}{' '}
```
Comment (kyranjamie) [inline src/components/stacking.tsx:61]: I had issues using this in wallet, doesn't handle big (unlikely but theoretically possible) numbers

File: src/pages/address/[principal].tsx [hunk 3]
```diff
@@ -26,19 +26,19 @@ import { TokenBalancesCard } from '@components/balances/principal-token-balances
...
 } from '@common/utils/addresses';
-import { StacksTokenVestingCard } from '@components/stacks-token-vesting-card';
 import { getNetworkMode } from '@common/api/network';
 import { useNetworkMode } from '@common/hooks/use-network-mode';
 import { IconAlertCircle } from '@tabler/icons';
 import { AddressLink } from '@components/links';
 import { useInfiniteFetch } from '@common/hooks/use-fetch-blocks';
 import { microStxToStx } from '@stacks/ui-utils';
+import { useUnlockingState } from '@common/hooks/use-unlocking-state';
+import { useEffect } from 'react';
```
Comment (kyranjamie) [inline src/pages/address/[principal].tsx:40]: Third party deps mixed with local code bothers me more than relative paths lol (though some do still slip the net)

I've tried tireless to get a good eslint import/order set up, but never had it perfect

File: src/pages/api/vesting/[address].ts [hunk 4]
```diff
@@ -1,19 +1,85 @@
...
+    return data;
+  } catch (e) {
+    throw Error(`Can't fetch actual balance: ${e.message}`);
+  }
+}
+
+async function fetchAverageBlockTime() {
+  const res = await fetch('https://blockchain.info/q/interval');
+  const data = await res.text();
+  return data;
```
Comment (kyranjamie) [inline src/pages/api/vesting/[address].ts:32]:
```suggestion
  return res.text();
```

File: src/pages/api/vesting/[address].ts [hunk 5]
```diff
@@ -1,19 +1,85 @@
...
+          fetchLegacyExplorerVestingData(result.mainnet.STACKS),
+          fetchAverageBlockTime(),
+        ]);
+
+        let lockedBalance = 0;
+        Object.keys(data).forEach((height, index) => {
+          if (parseInt(height) > currentHeight) {
+            lockedBalance += Object.values(data)[index];
+          }
+        });
```
Comment (kyranjamie) [inline src/pages/api/vesting/[address].ts:72]:
```suggestion
        const lockedBalance = Object.keys(data).reduce((acc, height) => {
          if (parseInt(height) > currentHeight) return acc + parseInt(height);
          return parseInt(height);
        }, 0);
```
can do more functionally something like ^ (untested)

**Review by kyranjamie — APPROVED** [review]

File: src/pages/api/vesting/[address].ts (hunk 5 above)
Comment (aulneau) [reply src/pages/api/vesting/[address].ts]: Thank you :)

**Comment by GinaAbrams** [comment]
the staging looks good to me! LMK if any other feedback would be helpful.

**Pull request merged**